vaadin-combo-box-flow/src/main/resources/META-INF/resources/frontend/comboBoxConnector.js

```diff
@@ -56,36 +79,59 @@ window.Vaadin.Flow.comboBoxConnector = {
 
       const filterChanged = params.filter !== lastFilter;
       if (filterChanged) {
-        pageCallbacks = {};
         cache = {};
         lastFilter = params.filter;
+        this._debouncer = Debouncer.debounce(
+          this._debouncer,
+          timeOut.after(500),
+          () => {
+            clearPageCallbacks();
+            comboBox.clearCache();
+            if (params.filter === '') {
+              // Fixes the case when the filter changes
+              // from '' to something else and back to ''
+              // within debounce timeout, and the
+              // DataCommunicator thinks it doesn't need to send data
+              comboBox.$server.resetDataCommunicator();
+            }
+          });
+        return;
       }
 
       if (cache[params.page]) {
         // This may happen after skipping pages by scrolling fast
         commitPage(params.page, callback);
       } else {
-        const upperLimit = params.pageSize * (params.page + 1);
-
-        if (filterChanged) {
+        pageCallbacks[params.page] = callback
+        const activePages = Object.keys(pageCallbacks).map(page => parseInt(page));
+        const rangeMin = Math.min(...activePages);
+        const rangeMax = Math.max(...activePages);
+
+        if (activePages.length * params.pageSize > MAX_RANGE_COUNT) {
+          if (params.page === rangeMin) {
+            clearPageCallbacks([String(rangeMax)]);
+          } else {
+            clearPageCallbacks([String(rangeMin)]);
+          }
+          comboBox.dataProvider(params, callback);
+        } else if (rangeMax - rangeMin + 1 !== activePages.length) {
+          // Wasn't a sequential page index, clear the cache so combo-box will request for new pages
+          clearPageCallbacks();
+        } else {
+          // The requested page was sequential, extend the requested range
+          const startIndex = params.pageSize * rangeMin;
+          const endIndex = params.pageSize * (rangeMax + 1);
+          const count = endIndex - startIndex;
+          setRequestedRange = () => comboBox.$server.setRequestedRange(startIndex, count, params.filter);
+          if (!this._debouncer || !this._debouncer.isActive()) {
+            setRequestedRange();
```

Comment:
This immeadiate initial request has a problem. Even if the combo-box should initially load e.g. 2 pages, it loads only the first one immediately and debounces the next request. The same can happen when you quickly scroll to a completely new scroll-position.

This could be fixed by wrapping this part in `setTimeout`, so that the immediately following requests would be included in the first batch. But for some reason, there's a noticeable delay between the `comboBox.dataProvider` calls, which means that the timeout should be several hundreds of milliseconds, which ruins the purpose.

This is not a huge issue, especially in most cases where a single page covers the viewport. We could create a further improvement ticket for this.

Reply:
Adding a timeout longer than a few milliseconds defeats the purpose of the initial call as you said so let's keep this as it is

Reply:
That's why it should be fixed by removing the delay between `dataProvider` calls. So that if combo-box needs two pages, it would immediately call `dataProvider` two times.

I will at least create a ticket so we don't lose the reference to this discussion.

Reply:
Nevermind. I think it was a noticable issue only when I was testing with really long debounce timeouts. With 200ms in the current implementation it's fine.